cltk/corpus/readers.py

```diff
@@ -207,3 +229,135 @@ def __iter__(self):
         """convenience iterator for Word2Vec training."""
         for sent in self.sents():
             yield sent
+
+
+class JsonfileCorpusReader(CorpusReader):
```

Comment:
A project history note: 3 years ago, we started generating these JSON files while working on a web UI and API (which have been put to the side, for now). In some cases we were coercing plaintext corpora into a JSON structure and in others were converting from XML. This class is definitely a nice-to-have but in the long run, we'll only really need this if someone upstream provides us with an "official" corpus already in JSON.

Reply:
I think the CLTK Json choice was a good one; and that it wasn't too much work to create the JsonfileCorpusReader to use it effectively, and although I initially didn't like it, I'm convinced that it is better than just the plaintext reader, which is only barely usable with the FilteredCorpusReader.  And yet both are fundamentally better than the mess that is the TEI files.

Reply:
> although I initially didn't like it, I'm convinced that it is better than just the plaintext reader

I have had a love-hate thing with the JSON texts. They're minimalistic yet capable of encoding any arbitrary text chunking and labels for those chunks (which can take the form or arabic numbers, roman numerals, stings, other numbers, etc etc), and handle infinite embedding.

Is it better than XML? It's a hell of a lot easier for programmers to pull from, though we do lose some information. I can think of two cases where we would need that "extra information":
* If we ever extended a reader to read from a fully marked up XML treebank, which contains (presumably) perfect human-made annotations. So in addition to getting word tokens, you could get attributes of each token like known-good lemma, etc. Keep in mind though that there are only so many XML treebank'ed classical texts out there.
* More important, for now at least, is that we will want I/O when building our machine learning models. There will be a fairly small number of users who would want to run their own models, however these are our most productive users and also most likely to contribute (hopefully) their high performing models back to us.

Reply:
Clarification: In the first bullet point, I am thinking of an assistant to human students of texts; in the latter, just pipeline for ML.

Reply:
Yes, I agree on your points.
I like Json for ease of use, but don't like the lack of schema validation, and it's not very human readable.
I like Xml for schema validation and flexibility, but I've only been disappointed in TEI and other efforts. I've hand written a lot of Xml and know that quality data structures are no easy task. Xml's original claim of human readability is of course questionable and debatable. 
Raw text files are of course, closest to "native format"--but lack of metadata organization puts all the library books on the floor. Also, programmatically accessing the data almost always brings a certain amount of dirtiness, as navigation often gets baked into the raw text data, etc.

Rather than pick a horse, I think we might be able to guide contributors and the external community at large by exposing metrics on corpus quality & promoting user-facing corpus selection utilities. I've started to do this in recent pull requests, but plan to elaborate more on this soon, most likely with some Jupyter notebooks, tutorials, articles and maybe a talk. Looking forward to feedback.

Reply:
XML and JSON aren't really comparable when used for their original purposes (inline document markup versus object serialisation) and it is unfortunate JSON wasn't around when Microsoft decided XML should be used for object serialisation as well as markup.

I don't think TEI is going away any time soon for archival purposes (or even when authoring markup). But, for straight object serialisation for APIs, JSON is generally better.

Of course you can combine the two: JSON for the data, XML for the marked-up documents.

For metadata, I find YAML the best from an authoring perspective (can always be transmitted as JSON in APIs).

But I do find flat files useful for storage and processing (again, potentially transmitting as JSON in APIs).

And there's also a lot to be said for stand-off markup.